saasherder/changelog.py

```diff
@@ -0,0 +1,158 @@
+import yaml
+from itertools import chain
+import os
+import subprocess
+
+
+class Changelog:
+
+    workspace = "_workspace/"
+
+    @staticmethod
+    def context_dir(context):
+        return context + "-services/"
+
+    def path(self, context, service_name):
+        """Get YAML config path for a service"""
+
+        return self.context_dir(context) + service_name + ".yaml"
+
+    def definitions(self, context, service_name):
+        """Get list of services defined in service file"""
+
+        with open(self.path(context, service_name)) as f:
+            parsed = yaml.load(f)
+            return parsed['services']
+
+    def service_names(self, context):
+        """Get names of all services in a context"""
+
+        dir = self.context_dir(context)
+        files = os.listdir(dir)
+
+        return [os.path.splitext(f)[0] for f in files]
+
+    def services(self, context):
+        """Get all services in the context"""
+
+        names = self.service_names(context)
+        dsls = [self.definitions(context, name) for name in names]
+        return list(chain(*dsls))
+
+    @staticmethod
+    def diff(s1, s2):
+        """Return the difference between 2 service lists
+
+        This is not a commutative relationship. We are looking for services in
+        first list which differ from that in second list. a ⊕ b ≠ b ⊕ a.
+
+        This gets messy when the hash refers to a branch or none. Ignoring such
+        services for now.
+        """
+
+        # All the services that changed
+        changed = [service for service in s1 if service not in s2]
+
+        def old(service):
+            cadidates = [s for s in s2 if s['name'] == service['name']]
+            return cadidates[0].get('hash') if cadidates else None
+
+        # Filter changed services for which I cannot get a changelog
+        useful = [s for s in changed if s.get('hash') and old(s)]
+
+        # Now changed services can be new ones or change in hash
+        return [{
+            'name': service['name'],
+            'url': service['url'],
+            'new': service.get('hash', 'master'),  # NOTE: null here is silly
+            'old': old(service)
+        } for service in useful]
+
+    @staticmethod
+    def run(cmd):
+        return subprocess.run(cmd, stdout=subprocess.PIPE,
```

Comment:
using this on `saas-openshift` repo makes it just sit there for a while - worth adding a `print ('Running ' + cmd)` to see progres.

Reply:
@maxandersen Yes, we need some more logging in general. I'll add that.